File: iosApp/iosApp/blockchain/events/commutoswap/SellerClosedEvent.swift
//
//  SellerClosedEvent.swift
//  iosApp
//
//  Created by jimmyt on 9/12/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import web3swift

/**
 Represents a [SellerClosed](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#sellerclosed) event, emitted by the CommutoSwap smart contract when the seller in a swap closes the swap.
 */
class SellerClosedEvent {
    /**
     The ID of the swap that the seller has closed.
     */
    let id: UUID
    /**
     The ID of the blockchain on which the event was emitted.
     */
    let chainID: BigUInt
    /**
     The hash of the transaction that emitted this event, as a lowercase hex string with a "0x" prefix.
     */
    let transactionHash: String
    /**
     Creates a new `SellerClosedEvent` given a web3swift `EventParserResultProtocol` containing information from a [SellerClosed](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#sellerclosed) event, or returns `nil` if the passed `EventParserResultProtocol` doesn't contain information from a SellerClosed event or does not have a valid transaction hash.
     */
    init?(_ result: EventParserResultProtocol, chainID: BigUInt) {
        guard let resultID = UUID.from(data: result.decodedResult["swapID"] as? Data) else {
            return nil
        }
        id = resultID
        self.chainID = chainID
        guard var transactionHash = result.transactionReceipt?.transactionHash.toHexString().lowercased() else {
            return nil
        }
        if !transactionHash.starts(with: "0x") {
            transactionHash = "0x" + transactionHash
        }
        self.transactionHash = transactionHash
    }
    /**
     Creates a new `SellerClosedEvent` with the given `id`, `chainID` and `transactionHash`.
     */
    init(id: UUID, chainID: BigUInt, transactionHash: String) {
        self.id = id
        self.chainID = chainID
        self.transactionHash = transactionHash
    }
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/blockchain/events/commutoswap/SellerClosedEvent.kt
package com.commuto.interfacemobile.android.blockchain.events.commutoswap

import com.commuto.interfacemobile.android.contractwrapper.CommutoSwap
import java.math.BigInteger
import java.nio.ByteBuffer
import java.util.*

/**
 * Represents a [SellerClosed](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#sellerclosed) event,
 * emitted by the CommutoSwap smart contract when the seller closes of a swap closes the swap.
 *
 * @property swapID The ID of the swap that the seller has closed.
 * @property chainID The ID of the blockchain on which this event was emitted.
 */
data class SellerClosedEvent(val swapID: UUID, val chainID: BigInteger) {

    companion object {
        /**
         * Creates a [SellerClosedEvent] from a [CommutoSwap.SellerClosedEventResponse] and a specified [chainID].
         *
         * @param event The [CommutoSwap.SellerClosedEventResponse] containing the ID of the swap as a [ByteArray].
         * @param chainID: The ID of the blockchain on which this event was emitted.
         *
         * @return A new [SellerClosedEvent] with a swap ID as a [UUID] derived from [event]'s swapID ID [ByteArray],
         * and the specified [chainID].
         */
        fun fromEventResponse(event: CommutoSwap.SellerClosedEventResponse, chainID: BigInteger): SellerClosedEvent {
            val swapIDByteBuffer = ByteBuffer.wrap(event.swapID)
            val mostSigBits = swapIDByteBuffer.long
            val leastSigBits = swapIDByteBuffer.long
            return SellerClosedEvent(UUID(mostSigBits, leastSigBits), chainID)
        }
    }
}